File: golang/set1/challenge8/main.go
package main

import (
	"bufio"
	"encoding/hex"
	"fmt"
	"os"
)

func getBlocks(ciphertext []byte, blockSize int) [][]byte {
	blocks := [][]byte{}
	for i := 0; i < len(ciphertext); i += blockSize {
		if i+blockSize > len(ciphertext) {
			blocks = append(blocks, ciphertext[i:])
		} else {
			blocks = append(blocks, ciphertext[i:i+blockSize])
		}
	}
	return blocks
}

func main() {
	blockSize := 16
	fh, err := os.Open("8.txt")
	if err != nil {
		panic(err)
	}
	reader := bufio.NewReader(fh)
	for {
		blockMap := make(map[[16]byte]bool)
		line, err := reader.ReadString('\n')
		if err != nil {
			return
		}
		truncatedLine := line[:len(line)-1]
		rawBytes, err := hex.DecodeString(truncatedLine)
		if err != nil {
			panic(err)
		}
		blocks := getBlocks(rawBytes, blockSize)
		//fmt.Printf("number of blocks %d\n", len(blocks))
		for _, block := range blocks {
			blockArr := [16]byte{}
			copy(blockArr[:], block)
			_, ok := blockMap[blockArr]
			if ok {
				fmt.Println("duplicate blocks detected!")
			} else {
				blockMap[blockArr] = true
			}
		}
	}
}

File: rust/set1/challenge8/src/main.rs

extern crate rustc_serialize as serialize;

use serialize::hex::FromHex;
use std::io::prelude::*;
use std::io::{BufReader};
use std::fs::File;
use std::collections::HashMap;

fn get_blocks<'a>(ciphertext: &'a [u8], block_size: u8) -> Vec<&'a [u8]> {
    let mut blocks: Vec<&'a [u8]> = vec![];
    let mut i = 0;
    for j in 0..ciphertext.len() {
        if i >= ciphertext.len() {
            break
        }
        if i + block_size as usize > ciphertext.len() {
            blocks.push(&ciphertext[i ..]);
        } else {
            blocks.push(&ciphertext[i .. i + block_size as usize]);
        }
        i += block_size as usize;
    }
    return blocks;
}

fn has_duplicate_blocks(ciphertext: &[u8], block_size: u8) -> bool {
    let mut block_map = HashMap::new();
    for block in get_blocks(&ciphertext, block_size) {
        if block_map.contains_key(block) {
            return true;
        } else {
            block_map.insert(block.to_owned(), true);
        }
    }
    return false;
}

fn main() {
    let fh = File::open("8.txt").unwrap();
    let fb = BufReader::new(fh);
    let key_size = 16;
    for line in fb.lines() {
        let ciphertext = line.unwrap().from_hex().unwrap();
        if has_duplicate_blocks(&ciphertext, key_size as u8) {
            println!("duplicate blocks detected");
        }
    }
}
